pytorch_lightning/callbacks/pt_callbacks.py

```diff
@@ -193,65 +197,91 @@ def __init__(self, filepath, monitor='val_loss', verbose=0,
+
+        # make paths
+        os.makedirs(dirpath, exist_ok=True)
 
         # delegate the saving to the model
         self.save_function(filepath)
 
+    def check_monitor_top_k(self, current):
+        return ((len(self.best_k_models.keys()) < self.save_top_k) or
+                (self.monitor_op(current, self.best_k_models[self.kth_value])))
+
     def on_epoch_end(self, epoch, logs=None):
         logs = logs or {}
-        self.epochs_since_last_save += 1
-        if self.epochs_since_last_save >= self.period:
-            self.epochs_since_last_save = 0
+        self.epochs_since_last_check += 1
+        if self.epochs_since_last_check >= self.period:
+            self.epochs_since_last_check = 0
             filepath = '{}/{}_ckpt_epoch_{}.ckpt'.format(self.filepath, self.prefix, epoch + 1)
-            if self.save_best_only:
+            if self.save_top_k:
                 current = logs.get(self.monitor)
                 if current is None:
                     print('Can save best model only with %s available,'
                           ' skipping.' % (self.monitor), RuntimeWarning)
                 else:
-                    if self.monitor_op(current, self.best):
+                    if (self.check_monitor_top_k(current)):
+                        if len(self.best_k_models.keys()) == self.save_top_k:
+                            # need to pop the kth
+                            delpath = '{}/{}_ckpt_epoch_{}.ckpt'.format(
+                                self.filepath, self.prefix, self.kth_value + 1)
+                            self.best_k_models.pop(self.kth_value)
+                            self._del_model(delpath)
+                        self.best_k_models[epoch] = current
+                        if len(self.best_k_models.keys()) == self.save_top_k:
+                            # monitor dict has reached k elements
+                            if self.mode == 'min':
+                                self.kth_value = max(self.best_k_models, key=self.best_k_models.get)
+                            else:
+                                self.kth_value = min(self.best_k_models, key=self.best_k_models.get)
+                        if self.mode == 'min':
+                            self.best = min(self.best_k_models.values())
+                        else:
+                            self.best = max(self.best_k_models.values())
                         if self.verbose > 0:
-                            print('\nEpoch %05d: %s improved from %0.5f to %0.5f,'
-                                  ' saving model to %s'
-                                  % (epoch + 1, self.monitor, self.best,
-                                     current, filepath))
-                        self.best = current
-                        self.save_model(filepath, overwrite=True)
+                            print('\nEpoch %05d: %s reached %s (best %s),'
+                                  ' saving model to %s as top %d'
```

Comment:
use f strings.   

a = f'{some_val} > {other_val}'

Reply:
The original code was using `%` instead of `f strings`, btw

Reply:
yeah, not great either haha. whenever we run into non f strings let's replace them